### DISCUSSION TITLE
Automating Uplinking Files Through GDS Integration API/GDS CLI

### DISCUSSION THREAD

**Discussion by mattoro** [nasa/fprime#1793] [Q&A]
Hello,

I am fairly new to working with fprime, but I am currently attempting to automate uplinking files at set intervals through the fprime GDS. I understand the GDS GUI separates an uplink into a staging process and an uplink process, and was wondering how this could be replicated and automated using the GDS Integration API, extending the GDS CLI, or if another method is more viable. I notice that "Adding CLI tools for file uplink/downlink" is listed as a new feature in the GDS CLI Developer's Guide, but was unable to find this feature in the src code.

Thank you

**Comment by LeStarch** [comment]
I do not believe that file uplink/downlink has been added to the test API.  However, you could write a script that does this.  The way I might approach this problem is to write a script that does the following:

1. Setup the standard pipeline stack as `fprime-gds` does
2. Use the components found here: https://github.com/fprime-community/fprime-gds/tree/devel/src/fprime_gds/common/files to uplink and downlink files as necessary

Please feel free to post follow-up questions.

**Reply by mattoro** [reply]
Thanks for the reply! 

I currently have a pytest environment which sets up the standard pipeline similar to ref_integration_test.py. I am currently unsure what the correct workflow is for queuing a file and uplinking using the FileUplinker component you specified in your response.

For instance, I am not sure if the correct process is using FileUplinker's enqueue() ->start() -> finish(), or is the FileQueue's run() method preferred? The goal is to upload a number of files from a specified filepath and uplink the entire batch of files, check for success, then repeat at a specified interval. If I need to clarify anything, please let me know!

Thanks
